src/haptic_buzz.c

```diff
@@ -0,0 +1,123 @@
+        return 0;
+    }
+    if (((state->sat > SAT_NONE) && (state->state == STATE_RUNNING))) {
+
+        if (state->sat == SAT_PB_DUTY) {
+            haptic_buzz->haptic_type = haptic_buzz->haptic_buzz_duty;
+        } else if (state->sat == SAT_PB_HIGH_VOLTAGE) {
+            haptic_buzz->haptic_type = haptic_buzz->haptic_buzz_hv;
+        } else if (state->sat == SAT_PB_LOW_VOLTAGE) {
+            haptic_buzz->haptic_type = haptic_buzz->haptic_buzz_lv;
+        } else if (state->sat == SAT_PB_TEMPERATURE) {
+            haptic_buzz->haptic_type = haptic_buzz->haptic_buzz_temp;
+        } else {
+            haptic_buzz->haptic_type = HAPTIC_BUZZ_NONE;
+        }
+
+        // This kicks it off till at least one ~300ms tone is completed
+        if (haptic_buzz->haptic_type != HAPTIC_BUZZ_NONE) {
+            haptic_buzz->haptic_tone_in_progress = true;
+        }
+    }
+
+    if (haptic_buzz->haptic_tone_in_progress || brake) {
+        haptic_buzz->haptic_counter += 1;
+
+        float buzz_current = fminf(20, haptic_buzz->haptic_buzz_intensity);
+        // small periods (1,2) produce audible tone, higher periods produce vibration
+        int buzz_period = haptic_buzz->haptic_type;
+        if (haptic_buzz->haptic_type == HAPTIC_BUZZ_ALTERNATING) {
+            buzz_period = 1;
+        }
+
+        // alternate frequencies, depending on "mode"
+        buzz_period += haptic_buzz->haptic_mode;
+
+        if (brake) {
+            // This is to emulate the equivalent of "stop click"
+            buzz_current = fmaxf(3, startup_click_current * 0.8);
+            buzz_current = fminf(10, buzz_current);
+            buzz_period = 0;
+        } else if ((abs_erpm < 10000) && (buzz_current > 5)) {
+            // scale high currents down to as low as 5A for lower erpms
+            buzz_current = fmaxf(haptic_buzz->haptic_buzz_min, abs_erpm / 10000 * buzz_current);
+        }
+
+        if (haptic_buzz->haptic_counter > buzz_period) {
+            haptic_buzz->haptic_counter = 0;
+        }
+
+        if (haptic_buzz->haptic_counter == 0) {
+            if (haptic_buzz->applied_haptic_current > 0) {
+                haptic_buzz->applied_haptic_current = -buzz_current;
+            } else {
+                haptic_buzz->applied_haptic_current = buzz_current;
+            }
+
+            if (fabsf(haptic_buzz->haptic_timer - current_time) > note_period) {
+                haptic_buzz->haptic_tone_in_progress = false;
+                if (brake) {
+                    haptic_buzz->haptic_mode += 1;
+                } else {
+                    if (haptic_buzz->haptic_type == HAPTIC_BUZZ_ALTERNATING) {
+                        haptic_buzz->haptic_mode = 5 - haptic_buzz->haptic_mode;
+                    } else {
+                        haptic_buzz->haptic_mode = 1 - haptic_buzz->haptic_mode;
+                    }
+                }
+
+                haptic_buzz->haptic_timer = current_time;
+            }
+        }
```

Comment:
IMO this piece of code should be factored out into an API similar to that of the `foc_play_tone()`, that is e.g. `buzz_tone(float frequency, float current)`. It'll be required to add the 6.05 `foc_play_tone()` as an alternative in a clean way.

Then, `haptic_buzz_update()` will prepare the data for the particular tone to play (according to the type of buzz) into the `hapitc_buzz` structure and `buzz_tone()` can modulate the buzz current onto the output current.

Reply:
> First of all IIRC haptic_tone is limited to specific frequency range and cannot play arbitrary frequencies

Oh, right, so maybe the vibrating haptics will have to remain via the current modulation?

> Also, with the current implementation, haptic tone stops playing when haptic_feedback_update isn't called in refloat_thd (as it's not being modulated onto the current anymore). This is different than foc_play_tone() and foc_stop_audio() which will start playing the tone once foc_play_tone() is called and foc_stop_audio() needs to be explicitly called to stop playing. i.e. That means the both methods have uniquely different requirements regarding state management in the finite state automata defined in refloat_thd.

I don't really see this as a big problem. You just output the `start_time`, `duration` and `type` of tone to play from `haptic_feedback_update()`. Then the application of the tone will be done in two separate places and slightly differently, but you got all the data you need.

For the alerts that are continuous in nature, like the pushbacks, you can keep the `duration` at say 300ms (which will give you a convenient minimum tone duration, though one that is applied even if the tone is long enough, but you can tweak that if you want) and continuously keep updating the `start_time` to keep the tone playing.

Reply:
mmkay, so I factored that stuff out into a new function and also put placeholder comments where I think 6.05 stuff makes sense to be implemented (either updating pid or calling foc_play_tone)